### PR TITLE
Depth sorting for node faces

### PR DESCRIPTION
**Pull request by x2048**
This PR aims to implement correct painter's algorithm for transparent nodes on the map.

* Vertex indexes for all transparent faces are captured during mesh generation.
* [BSP Tree](https://en.wikipedia.org/wiki/Binary_space_partitioning) is generated for transparent triangles.
* When rendering the map, BSP tree is traversed to generate correct (in most cases) rendering order, which is applied by replacing indexes in the mesh buffers as suggested by @hecktest.

This PR supersedes parts of #11130 and fixes most of cases of transparency problems (including transparent nodebox nodes at the cost of performance).

## Handling Transparent Nodeboxes

In order to render overlapping nodeboxes correctly, this PR splits each nodebox at the edges of other nodeboxes along the same axis. This process produces a significantly larger number of 'atomic' nodeboxes that are guaranteed not to overlap, and may impact overall rendering performance. The total number of nodeboxes (defined + produced) is currently limited to 100.

## To do

This PR is Ready for Review and testing

## How to test

Use a combination of semi-transparent nodes such as:
* Alpha testing nodes in devtest
* Water
* Colored glass in Mineclone 2
* Stained glass 

Transparent nodes should be rendered correctly with minimal temporary artifacts from all angles.

### CODE REVIEW

**Comment by x2048** [comment]
Tagging @appgurueu @Wuzzy2 @sfan5 @Andrey2470T @hecktest

**Comment by Andrey2470T** [comment]
Seems it has fixed the main transparency issue, transparent nodes are rendered behind other transparent ones now... but:
screenshot_20211016_235712

I used the colored glass from `glass_stained` mod and it caused that artifact, some parts of mesh faces of the nodes are black and those, as I understood, that are beyond the node area. Here's how it looks like in minetest without the bugfix:
screenshot_20211016_235641

**Comment by Andrey2470T** [comment]
Now after the fix FPS has become higher on that server, even a bit higher 8 (varies within 10 - 13 FPS). Thanks for the fix.

**Comment by Andrey2470T** [comment]
I've discovered a wonderful bug with this PR (server `Tunnelers Abyss`):
Снимок экрана от 2021-11-04 23-34-20

Снимок экрана от 2021-11-04 23-37-10

Снимок экрана от 2021-11-04 23-37-31

Looks like as if somebody strongly extruded the colored glass panes to the outside.

**Comment by Andrey2470T** [comment]
So is this bug irrelevant to this PR if you told me you didn't discover this on Tunneler's Abyss? I don't see its display on any another builds, but only in your one. I doubt this problem concerns to my hardware or video drivers. So, I can conclude this bug hides somewhere in your PR's code.

**Comment by x2048** [comment]
I can confirm the bug, but it does not occur in 100% of cases.

**Comment by x2048** [comment]
@Andrey2470T The bug was caused by an uninitialized memory read due to reallocation inside std::vector and the reference to the current box not being adjusted. It should be fixed now, and I've also added a few improvements to the code to avoid performance problems in edge cases.

**Comment by appgurueu** [comment]
Could the BSP tree be made to take into account semitransparent particle & entity triangles?

**Comment by x2048** [comment]
I would not say it is impossible, but it is a significant step up in complexity. There are multiple reasons for this:
* Current BSP tree is per mapblock and is built on assumptions of a mapblock mesh (e.g. axis-aligned cubes). Building a generic BSP tree for free-form objects requires some good math and mesh manipulation on the fly (adding new vertices).
* BSP trees in general are expensive to build and are usually used for static meshes (like the map blocks). Putting a moving object into a BSP tree means rebuilding the tree on each frame, and it's at least O(N logN) on the number of triangles.
* Our current rendering process simply not fit for this at the moment, we would need to extract a deferred transparent rendering pass, and I am not sure Irrlicht is helpful here.

**Comment by Andrey2470T** [comment]
Is this PR ready for merging? In any case, I didn't found any other noticable bugs, except which I reported about before and which already fixed. The only current downside is a lower FPS, however generally it is not critical and quite playable.

**Comment by x2048** [comment]
This PR is ready to be reviewed by the core developers, there is no known technical debt / problems to be fixed.

**Comment by erlehmann** [comment]
@x2048 please rebase this branch on master so it can be compared to recent Minetest easier in terms of performance.

File: src/client/clientmap.cpp
Comment (erlehmann) [inline src/client/clientmap.cpp:417]: Is the added whitespace here really necessary?

File: src/client/clientmap.cpp
Comment (erlehmann) [inline src/client/clientmap.cpp:817]: What kind of scenario triggers this?

**Comment by runsy** [comment]
Yes. I have issues:

[](https://forum.minetest.net/download/file.php?id=25605)

Transparent nodes besides othe rtransparent nodes.

**Comment by x2048** [comment]
@runsy Can you give more detail on how to reproduce it?

**Comment by wsor4035** [comment]
@x2048 seems it was taken from https://forum.minetest.net/viewtopic.php?p=406861#p406861 which appears to be regular mte and not this PR? follow up https://forum.minetest.net/viewtopic.php?p=407177#p407177

**Comment by x2048** [comment]
@runsy The problem on your screenshot is fixed by this PR.

@wsor4035 Thank you for the pointers.

**Comment by runsy** [comment]
> @x2048 seems it was taken from https://forum.minetest.net/viewtopic.php?p=406861#p406861 which appears to be regular mte and not this PR? follow up https://forum.minetest.net/viewtopic.php?p=407177#p407177

I expect my issue solved with this PR. I do not check it.

Ca n anyone provide me this PR in ZIP to compile it in order to test it?

**Comment by runsy** [comment]
> @runsy The problem on your screenshot is fixed by this PR.
> 
> @wsor4035 Thank you for the pointers.

Cool. So when merged? I expect this in 5.6.

File: src/client/clientmap.cpp
Comment (sfan5) [inline src/client/clientmap.cpp:103]: general advice: vectors are only a couple bytes so using a constref for them doesn't make much sense

File: src/client/clientmap.cpp
Comment (sfan5) [inline src/client/clientmap.cpp:352]:
```suggestion
		updateTransparentMeshBuffers();
```

File: src/client/clientmap.cpp
Comment (sfan5) [inline src/client/clientmap.cpp:792]: Could use a reserve call *before* the loop, this probably works:
`draw_order.reserve(draw_order.size() + list.bufs.size());`

File: src/client/clientmap.cpp
Comment (sfan5) [inline src/client/clientmap.cpp:954]: this might be clearer written as `powf(75 * BS, 2)`

File: src/client/content_mapblock.cpp
Comment (sfan5) [inline src/client/content_mapblock.cpp:385]: If this introduces a new requirement that origin must be added *after* calling `drawNodeboxNode` that should be written down where `drawNodeboxNode` is defined or so.

File: src/client/content_mapblock.cpp
Comment (sfan5) [inline src/client/content_mapblock.cpp:1406]: why stop at 100?

File: src/client/content_mapblock.cpp
Comment (sfan5) [inline src/client/content_mapblock.cpp:1393]: this can be optimized regarding allocations like so:
```suggestion
		std::vector<float> sections;
		sections.reserve(8 + 2 * boxes.size());
		for (int axis = 0; axis < 3; axis++) {
			// identify sections
			sections.clear();
```

File: src/client/mapblock_mesh.cpp
Comment (sfan5) [inline src/client/mapblock_mesh.cpp:1050]: generally you can replace all `.size() == 0` or `.size() > 0` here with `.empty()` or `!.empty()` respectively

File: src/client/mapblock_mesh.cpp
Comment (sfan5) [inline src/client/mapblock_mesh.cpp:1017]:
```suggestion
	std::vector<s32> indexes;
	indexes.reserve(triangles->size());
```

File: src/client/mapblock_mesh.cpp
Comment (sfan5) [inline src/client/mapblock_mesh.cpp:1061]: create a constructor for this?

File: src/client/mapblock_mesh.cpp
Comment (sfan5) [inline src/client/mapblock_mesh.cpp:1348]: wasn't `isTransparent()` intended for this?

File: src/client/mapblock_mesh.cpp
Comment (sfan5) [inline src/client/mapblock_mesh.cpp:1517]:
```suggestion
			if (current_buffer) {
				m_transparent_buffers.emplace_back(current_buffer, current_strain);
```

File: src/client/mapblock_mesh.cpp
Comment (sfan5) [inline src/client/mapblock_mesh.cpp:1529]:
```suggestion
	if (!current_strain.empty())
		m_transparent_buffers.emplace_back(current_buffer, current_strain);
```

File: src/client/mapblock_mesh.h
Comment (sfan5) [inline src/client/mapblock_mesh.h:149]: I don't remember seeing any code that'd require this.

File: src/client/mapblock_mesh.h
Comment (sfan5) [inline src/client/mapblock_mesh.h:211]: could be const

File: src/nodedef.cpp
Comment (sfan5) [inline src/nodedef.cpp:669]: looks unused

File: src/client/tile.h
Comment (sfan5) [inline src/client/tile.h:262]: looks unused

**Comment by runsy** [comment]
@sfan5 Please review this one.

**Review by sfan5 — CHANGES_REQUESTED** [review]
This review was mostly reading code though I gave it a try ingame and tested a few random things, it was fine.

**Comment by sfan5** [comment]
Question: Is this whole ordeal necessary for clipped alpha (no blending)? The code seems to do it for those (or at least `isTransparent` checks them too...).

**Comment by appgurueu** [comment]
> Question: Is this whole ordeal necessary for clipped alpha (no blending)? The code seems to do it for those (or at least `isTransparent` checks them too...).

Not if we have a `discard` in our shaders.

File: src/client/clientmap.cpp
Comment (x2048) [reply src/client/clientmap.cpp:817]: Looks like going under roughly 1fps will trigger this condition.

File: src/client/content_mapblock.cpp
Comment (x2048) [reply src/client/content_mapblock.cpp:1408]: This is a silly cap to prevent crashing away in case the inner loop starts adding an infinite number of new boxes due to e.g. precision glitches.

Number 100 is an arbitrary, reasonably high number that seems to work with the nodes defined today.

Comment (sfan5) [reply src/client/content_mapblock.cpp:1408]: That's okay, just please write it down in a comment.

**Comment by lhofhansl** [comment]
Part of this is already merged. Do we still need this one? @x2048

**Comment by x2048** [comment]
@lhofhansl Yes we do. We merge per-mapblock sorting, this adds per-face sorting.

**Comment by kilbith** [comment]
@x2048 [PTAL](https://irc.minetest.net/minetest-dev/2022-02-23#i_5943075).

**Comment by x2048** [comment]
@kilbith OIT is a game changer.
@appgurueu  too bad I did not see https://jcgt.org/published/0002/02/09/ earlier.

@sfan5 @lhofhansl would you still like to see this merged, until order-independent transparency is implemented?

File: src/client/content_mapblock.cpp
Comment (x2048) [reply src/client/content_mapblock.cpp:385]: Origin is added in line 388 below, I just moved generation of texture coords earlier, because having origin added breaks it.

File: src/client/tile.h
Comment (x2048) [reply src/client/tile.h]: Removed

File: src/nodedef.cpp
Comment (x2048) [reply src/nodedef.cpp]: Removed

File: src/client/mapblock_mesh.h
Comment (x2048) [reply src/client/mapblock_mesh.h]: It's now const along with a bunch of other methods.

File: src/client/mapblock_mesh.h
Comment (x2048) [reply src/client/mapblock_mesh.h]: Unfriended

File: src/client/content_mapblock.cpp
Comment (x2048) [reply src/client/content_mapblock.cpp:385]: It does not. I merely moved generation of texture coordinates earlier because it was broken by adding origin too early.

File: src/client/clientmap.cpp
Comment (x2048) [reply src/client/clientmap.cpp:792]: I decided to preallocate draw_order once, since there are a lot of buffers, and moving memory feels more expensive.

File: src/client/clientmap.cpp
Comment (x2048) [reply src/client/clientmap.cpp:103]: I was hoping that it would allow passing arguments in registers, rather than pushing on stack.

File: src/client/mapblock_mesh.cpp
Comment (x2048) [reply src/client/mapblock_mesh.cpp]: replaced

File: src/client/content_mapblock.cpp
Comment (x2048) [reply src/client/content_mapblock.cpp:1411]: done

**Comment by SmallJoker** [comment]
What's the best approach to compare the performance of this PR vs master? Is it sufficient to compare the `renderMap` F6 profiler entries? Do I need mapblocks filled with transparent nodes (glass?) to trigger the new code?

> would you still like to see this merged, until order-independent transparency is implemented?

What would your plans be in this case? I cannot tell what would be better in this case due to lack of knowledge in the rendering code, but having a general approach for all rendered parts (objects, nodes) would be really compelling to keep a simple/clear code structure.

File: src/client/clientmap.cpp
Comment (sfan5) [reply src/client/clientmap.cpp]: While a pointer definitely fits in a register I think having to dereference it is still worse than reading everything close together from the stack.
Anyway, there are still a few more `const v3s16 &` and `const v3f &`

**Comment by sfan5** [comment]
> @sfan5 @lhofhansl would you still like to see this merged, until order-independent transparency is implemented?

sure

**Comment by x2048** [comment]
@SmallJoker I created a pattern out of transparent nodes in devtest to trigger the worst-case scenario:
image
The height of each column is 44 nodes.

This case renders at 60 FPS without depth sorting and at 10-12 FPS with depth sorting on my laptop (Ryzen 7), so there is definitely a performance penalty associated.

**Comment by runsy** [comment]
> @SmallJoker I created a pattern out of transparent nodes in devtest to trigger the worst-case scenario: image The height of each column is 44 nodes.
> 
> This case renders at 60 FPS without depth sorting and at 10-12 FPS with depth sorting on my laptop (Ryzen 7), so there is definitely a performance penalty associated.

Too much FPS drop. HUGE. REALLY ANNOYING.

Any better solution?

My ideas:

1. Only transparency til a depth (maybe 3 - 5 nodes)
2. An optional feature. Disabled by default.
3. A better algorithm. Study the Minecraft / other games approach.

**Comment by x2048** [comment]
@runsy Does it mean that you suggest closing this PR without merging it in?

I explicitly showed the worst case with 22x22x44 nodes with 4 materials and lots of overlaps. There is a limit of 75 nodes for transparency sorting, about five blocks from the player, and I can of course make it configurable.

OIT has a good chance to be faster but will increase average load on the GPU, so not everyone will benefit from it. This is a topic for a different PR.

**Comment by Wuzzy2** [comment]
How about adding a setting which sets a maximum as to how much complexity is allowed, to allow fine-tuning if neccessary. That way, players with weak computers can turn it off or set to a low value.

**Comment by runsy** [comment]
> @runsy Does it mean that you suggest closing this PR without merging it in?

What I'm saying is that since it makes frame drops of 80% with respect to normal, make it optional. Or as sfan5 says tuneable.

**Comment by appgurueu** [comment]
> > @runsy Does it mean that you suggest closing this PR without merging it in?
> 
> What I'm saying is that since it makes frame drops of 80% with respect to normal (60FPS -> 10 FPS is too much, a frame drop of 5-10FPS would be reasonable), make it optional. Or as wuzzy says tuneable.

in the *worst case*

**Comment by runsy** [comment]
> > > @runsy Does it mean that you suggest closing this PR without merging it in?
> > 
> > 
> > What I'm saying is that since it makes frame drops of 80% with respect to normal (60FPS -> 10 FPS is too much, a frame drop of 5-10FPS would be reasonable), make it optional. Or as wuzzy says tuneable.
> 
> in the _worst case_

I always heard that minetets supported old equipment and blah blah blah. Let's keep it real.

**Comment by appgurueu** [comment]
> > in the _worst case_
> 
> I always heard that minetets supported old equipment and blah blah blah. Let's keep it real.

You're the one not keeping it real by arguing using the worst case (the only valid point I could see is griefers abusing this to lower FPS of other players in certain areas). If semitransparent nodes are randomly distributed, this worst case definitely won't come into play.

To give another example: Quicksort has a worst-case complexity of O(n²) as well, yet it is widely used, because the expected average running time if randomly picking pivots is O(n log n). Granted, in this case the complexity isn't determined by the *data* (the list) being "average", but rather the random not constantly picking the worst pivots.

**Comment by runsy** [comment]
So, what is the final decision?

It is necessary another dev to review this.

**Comment by x2048** [comment]
@runsy The distance of depth sorting will become adjustable, which will allow the client to avoid performance drops. The PR goes in, and we start work on OIT that would support both map, entities, clouds and particles. And be less sensitive to edge cases.

**Comment by runsy** [comment]
> @runsy The distance of depth sorting will become adjustable, which will allow the client to avoid performance drops. The PR goes in, and we start work on OIT that would support both map, entities, clouds and particles. And be less sensitive to edge cases.

coolz

**Comment by appgurueu** [comment]
> coolz

Please use a reaction instead of a message for this. Related: Could we use some guidelines for discussion on GH?

**Comment by sfan5** [comment]
@x2048 if you want this merged there's only two open comments from my side.

**Comment by SmallJoker** [comment]
grafik

Basically instant OOM while facing transparent nodeboxes near the spawn of Blocky Survival (mt.krypticbit.com 30005). This definitely need investigation.

**Comment by x2048** [comment]
@SmallJoker good stuff, thanks for sharing

**Comment by x2048** [comment]
@SmallJoker OOM should be fixed now.

**Comment by x2048** [comment]
@runsy i've now added a setting `transparency_sorting_distance` to control the performance impact of transparency sorting.

File: src/client/mapblock_mesh.cpp
Comment (x2048) [reply src/client/mapblock_mesh.cpp]: Done

**Comment by runsy** [comment]
> @runsy i've now added a setting `transparency_sorting_distance` to control the performance impact of transparency sorting.

Coolz Thankz.

File: src/client/clientmap.cpp
Comment (x2048) [reply src/client/clientmap.cpp]: Done

File: src/client/content_mapblock.cpp
Comment (SmallJoker) [inline src/client/content_mapblock.cpp:1399]: Might it be more efficient if you preload `sections` with these values outside of the loop and use `sections.resize(n);`?
However I do not know whether this affects the capacity changed by `sections.reserve` above.

File: src/client/mapblock_mesh.cpp
Comment (SmallJoker) [inline src/client/mapblock_mesh.cpp:1027]: Would you please be so nice to add a description to this function? It would be nice to at least get an idea of how this function helps the depth sorting. More helpful comments = easier to understand for maintenance.

File: src/client/mapblock_mesh.cpp
Comment (SmallJoker) [inline src/client/mapblock_mesh.cpp:1040]:
```C++
for (const MeshTriangle &t : list) {
	if (t.areaSQ > candidate_triangle->areaSQ ||
```
The range-based `for` loop is shorter, supersedes an additional variable and `t` still allows address-of operators for the return value.

**Review by SmallJoker — APPROVED** [review]
Works. Code adjustments are welcome.

File: src/client/mapblock_mesh.cpp
Comment (x2048) [reply src/client/mapblock_mesh.cpp:1040]: Well, `list` contains indexes of the triangles, and I get the actual triangle from `triangles`. I don't think I an do this in a different way.

File: src/client/content_mapblock.cpp
Comment (x2048) [reply src/client/content_mapblock.cpp]: done

File: src/client/mapblock_mesh.cpp
Comment (x2048) [reply src/client/mapblock_mesh.cpp:1036]: Done for this function and added a link to the method to BspTree class.

File: src/client/mapblock_mesh.cpp
Comment (SmallJoker) [reply src/client/mapblock_mesh.cpp:1049]: Oh yes, you are totally right. :+1:

**Review by sfan5 — APPROVED** [review]
Gave it a quick test, I don't have any intensive test scenes so I trust @SmallJoker with that.

**Pull request merged**
